Adding RRef as return value for builtin operators

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25499 Add Python RRef as args and return value
* **#25169 Adding RRef as return value for builtin operators**

See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Adding RRef as return value for builtin operators

See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "[WIP] Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

zhaojuanmao commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -174,7 +175,7 @@ std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
   auto requestId = nextId();
   auto future = std::make_shared<FutureMessage>();
-  if (message.isRequest()) {
+  if (message.requiresResponse()) {
```
how does it work when isRequest() is superset of requiresResponse() while  isRequest() is inside if(requiresResponse())?

mrshenli replied:
Sorry, didn't follow. Where is the inner `isRequest`?

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,107 @@
+  static std::unique_ptr<RRefContext>& getInstance();
+
+  RRefContext(const RRefContext &) = delete;
+  void operator=(const RRefContext &) = delete;
```
It seems delete is more explicit? e.g., [this post](https://stackoverflow.com/questions/18847957/delete-modifier-vs-declaring-function-as-private)

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,107 @@
+    if (ownerId == getWorkerId()) {
+      return getOrCreateOwnerRRef<T>(genRRefId());
+    } else {
+      return createUserRRef<T>(ownerId, genRRefId(), genRRefId());
```
For `UserRRef`, we need to call `genRRefId()` twice to get different ids. Because user RRef's `rrefId` and `forkId` should be different. 

But, yes I should've moved `forkId` from `RRef` to `UserRRef`, as `UserRRef` don't need to have a fork id.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -0,0 +1,69 @@
+std::unique_ptr<RRefContext>& RRefContext::getInstance() {
+  TORCH_CHECK(RRefContext::context_,
+      "Have to initialize RRefContext before use.");
+  return RRefContext::context_;
```
I thought about this initially, but not sure how to achieve that. `context_` needs to be static, and hence needs to be initialized in cpp file before the `RRefContext::init(agent_)` is called. However, initializing `context_` does require a ptr to `agent_`, which is not available when the class is loaded. A `unique_ptr` is a work around of that. Any suggestion?

mrshenli replied:
that means users can call `getInstance()` to retrieve the static `RRefContext` without setting the `agent`, and we have to check `isInitialized` on every method call on `RRefContext`?

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -70,6 +70,36 @@ std::shared_ptr<FutureMessage> py_rpc_python_udf(
                             MessageType::PYTHON_CALL));
 }
 
+std::shared_ptr<RRef> py_remote_builtin(
```
ahh, yes, forgot to clean this up.

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

pietern commented on test/test_rpc.py:
```diff
@@ -378,6 +379,44 @@ def test_stress_light_rpc(self):
+        values = []
+        t = threading.Thread(target=all_to_here, args=(rrefs, values))
+        t.start()
+        t.join(int(get_timeout(self.id()) / 2))
```
Why run this on a separate thread?

mrshenli replied:
I was a little worried it might take too long to finish all `to_here` calls. At a second thought, the default timeout should sufficient. I will remove this thread.

pietern commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +59,47 @@ Message processRequestBlocking(Message&& request) {
+      auto ownerRRef = ctx->getOrCreateOwnerRRef<IValue>(rrefId);
+
+      // TODO: make this asynchronous
+      auto stack = src.stack();
```
Does this need to be a copy? Looks like `src` has a short life time, so could be mutated in place with `auto&`.

mrshenli replied:
added a `stackRef` API to make it explicit.

pietern commented on torch/csrc/distributed/rpc/script_remote_call.cpp:
```diff
@@ -0,0 +1,62 @@
+      retRRefId_(std::move(retRRefId)),
+      retForkId_(std::move(retForkId)) {}
+
+at::IValue ScriptRemoteCall::retRRefId() {
```
`const &`?

pietern commented on torch/csrc/distributed/rpc/script_remote_call.cpp:
```diff
@@ -0,0 +1,62 @@
+  return retRRefId_;
+}
+
+at::IValue ScriptRemoteCall::retForkId() {
```
`const &`?

pietern commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -31,30 +66,46 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
+            op,
+            std::move(stack),
+            ownerRRef->id().toIValue(),
+            ownerRRef->id().toIValue()
```
This means owner rref == fork rref. The message handler defined above asserts they are not equal.

mrshenli replied:
this was intended for sending to self, which is not supported yet. Let me remove this for now.

pietern commented on torch/csrc/distributed/rpc/script_ret.cpp:
```diff
@@ -29,7 +29,8 @@ Message ScriptRet::toMessage() {
   auto payload_size = message.payload().size();
-  auto value = jit::unpickle(payload, payload_size, nullptr, &message.tensors());
+  auto value = jit::unpickle(
+      payload, payload_size, nullptr, &message.tensors());
```
This will conflict with #25502.

pietern reviewed (commented):
This is a first pass, need to do another one.

I noticed this is not clang-format'ed. Do you want to do that as part of this PR as well?

pietern commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +59,47 @@ Message processRequestBlocking(Message&& request) {
+      TORCH_CHECK(rrefId != forkId, "Does not support remote call to self.");
+
+      auto& ctx = RRefContext::getInstance();
+      auto ownerRRef = ctx->getOrCreateOwnerRRef<IValue>(rrefId);
```
Why is this created here but not passed the operation? I would assume that the rref that captures the result of the operation is something that _could_ be forwarded. If it isn't (yet, or ever), then the owner rref creation can be delayed until after the operation has completed successfully.

mrshenli replied:
In this PR, it can be moved to after the operation has completed. In the next one, it will do the following:

1. create an owner RRef
2. record the user RRef (the RRef on the caller)
3. send an `RREF_USER_ACCEPT` message back to the caller, so that the caller may delete it's local RRef after that. 
4. run the operation.

Hence, it needs the owner RRef before the operation execution. 

Let me move it to when the operation is done in this PR.

pietern commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -31,30 +66,46 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
-        Stack stack = torch::jit::createStackForSchema(
-            op->schema(), args, kwargs, c10::nullopt);
+  Stack stack;
+  auto op = match_builtin_op(opName, args, kwargs, stack);
```
Why snake case here and camel case for other function names?

pietern commented on torch/csrc/distributed/rpc/rref.cpp:
```diff
@@ -0,0 +1,118 @@
+          ScriptRRefFetch(id().toIValue()).toMessage()
+      );
+  auto srv = ScriptRRefValue::fromMessage(fm->wait());
+  return srv.value();
```
Is it possible to enforce move semantics here? Not sure if this will trigger a copy.

mrshenli replied:
Adding move here gives me "warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]". 

I modified the return type of `srv.value()` to `const&` to explicitly avoid additional copy.

pietern replied:
It is possible that returning as value implies move semantics here, given that `srv` is also local.

pietern replied:
A ref will be invalid after return, since srv is local.

mrshenli replied:
> A ref will be invalid after return, since srv is local.

Oh, I mean `srv.value()`'s return type is `IValue&`, but `toHere()`'s return type is still `IValue`. So that there will only be one copy.

pietern commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,137 @@
+  const worker_id_t ownerId_;
+  const RRefId rrefId_;
+  const ForkId forkId_;
+};
```
If this is sent over the wire as a buffer, without going through serialization, then it pays off to add:

```cpp
static_assert(std::is_trivially_copyable<RRefForkData>::value, "RRefForkData must be trivially copyable");
```

mrshenli replied:
It seems we need GCC version >= 5 to use `is_trivially_copyable`?

pietern replied:
Hmm crap... You're right.

But we have the following macro instead: https://github.com/pytorch/pytorch/blob/d8314a6260dd8c6de1d8e32d18e138d2b2d64cc1/c10/macros/Macros.h#L213-L217

mrshenli replied:
Let me use that.

pietern commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,137 @@
+
+  c10::optional<T> value_;
+  mutable std::mutex mutex_;
+  mutable std::condition_variable valueCV_;
```
Why are these marked as `mutable`?

mrshenli replied:
I would like to make `getValue()` function const, but it modifies the mutex and CV. What's the convention here? Should I drop the const modifier on `getValue()` instead?

pietern replied:
I see, I think it's fine then. @smessmer @ezyang do you have thoughts on this pattern?

ezyang replied:
I follow the standard C++ guidance, which is that 'const'-ness refers to logical constness, and thus mutable is an appropriate use for things like mutexes.

pietern commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,137 @@
+    AT_ERROR("OwnerRRef does not support toHere(), use getValue() instead.");
+  }
+
+  // TODO: add setValue(py::object) and getPyObj() for Python UDF
```
Comment should go.

pietern commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,110 @@
+      // NB: cannot use make_shared here as the constructor of OwnerRRef is
+      // private.
+      auto rref = std::shared_ptr<OwnerRRef<T>>(
+          new OwnerRRef<T>(getWorkerId(), rrefId));
```
Can use `std::make_shared` here.

mrshenli replied:
`OwnerRRef`'s constructor is private (this is intentional, so that all RRef construction goes through `RRefContext`). If we use make shared, `OwnerRRef` needs to add the allocator as a friend or we need to create a wrapper class that enables `make_shared`, no?

pietern reviewed (commented):
Another pass. Looking good!

aazzolini commented on torch/csrc/distributed/rpc/types.cpp:
```diff
@@ -0,0 +1,48 @@
+}
+
+GloballyUniqueId GloballyUniqueId::fromIValue(at::IValue&& ivalue) {
+  auto ivalues = ivalue.toTuple()->elements();
```
what happens if ivalue.toTuple() conversion fails? Does it throw an exception or does it return a nullptr?

mrshenli replied:
It throws an exception if it is not a tuple

aazzolini commented on torch/csrc/distributed/rpc/script_rref_proto.h:
```diff
@@ -0,0 +1,68 @@
+  static ScriptRRefCreate fromMessage(const Message& message);
+};
+
+// UserRRef (regardless of it's the creator or not) uses this message to notiify
```
typo: s/notiify/notify

aazzolini commented on torch/csrc/distributed/rpc/script_rref_proto.h:
```diff
@@ -0,0 +1,68 @@
+namespace rpc {
+
+// Temporary solution of RRef operations.
+// TODO: Remove all these messages and use rpc + registered functions instead.
```
What registered functions? You mean JIT built-ins?

mrshenli replied:
I think RRef should be built on top of RPC not in RPC. It does not has to be JIT function, we could have our own namespace for RRef. Similar to using `torch.ops.aten.*` for builtin ops, we could have `torch.distributed.rpc.*` for RPC related functions. Then all these messages can be unified as a `SCRIPT_CALL`.

aazzolini commented on torch/csrc/distributed/rpc/message.h:
```diff
@@ -12,6 +12,11 @@ enum MessageType {
   PYTHON_CALL,
   PYTHON_RET,
+  REMOTE_CALL,
+  RREF_FETCH,
```
nit: RREF_FETCH_CALL, RREF_FETCH_RET for naming consistency

aazzolini commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -257,9 +258,10 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
-      if (message.isRequest()) {
-        auto response = cb_(std::move(message));
-        send(work.from_, std::move(response));
+      if (message.requiresResponse()) {
```
I believe we should always send the response, at least as an acknowledgment, and the client should not throw away the acknowledgement, at least for tracing purposes. (e.g. if a FORK_DELETE times out, we may want to re-send the request. etc).

mrshenli replied:
Will address this in #25499

aazzolini reviewed (approved):
This looks good, but it seems that the messaging mechanism will have to change in order to accommodate  the remaining use cases, in particular allowing RRefs as arguments to built-ins. I believe we should commit this one soon and make the required changes on top.

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli committed:
Update on "Adding RRef as return value for builtin operators"


See #23110 for RRef design details. This commit only implements
RRef as return value for builtin operators, and RRef will communicate
between a user and the owner. More specifically, a RRef is first
created on the `dist.remote` caller, which is a user of the RRef.
Then the RRef user sends and notification to the owner to report
the fork to the owner, and the owner uses a shared_ptr to keep
the RRef alive. When the user RRef is destructed on the caller,
another notification will be sent to the owner, and the owner
can then drop it's RRef as well.

Differential Revision: [D17048343](https://our.internmc.facebook.com/intern/diff/D17048343)

mrshenli commented:
As of the latest commit in #25502, the conflicts no longer exist. Landing now.

pritamdamania87 commented on torch/csrc/distributed/rpc/script_remote_call.cpp:
```diff
@@ -0,0 +1,60 @@
+      retForkId_(std::move(retForkId)) {}
+
+const at::IValue& ScriptRemoteCall::retRRefId() {
+  return retRRefId_;
```
what does the **ret** in `retRRefId_` refer to?

pritamdamania87 commented on torch/csrc/distributed/rpc/script_remote_call.h:
```diff
@@ -0,0 +1,38 @@
+  static ScriptRemoteCall fromMessage(const Message& message);
+
+ private:
+  const at::IValue retRRefId_;
```
We should use ints for these as class member variables, IValue should only be used for serialization.

pritamdamania87 commented on torch/csrc/distributed/rpc/script_rref_proto.cpp:
```diff
@@ -0,0 +1,56 @@
+      jit::unpickle(payload, payload_size, nullptr, &message.tensors());
+  auto values = value.toTuple()->elements();
+
+  AT_ASSERT(values.size() == 1, "Expect a single IValue from message.");
```
TORCH_INTERNAL_ASSERT?

pritamdamania87 commented on torch/csrc/distributed/rpc/script_rref_proto.h:
```diff
@@ -0,0 +1,70 @@
+// UserRRef uses this message to fetch the remote RRef value from the owner.
+class TORCH_API ScriptRRefFetchCall final : public RRefMessageBase {
+ public:
+  ScriptRRefFetchCall(at::IValue rrefForkData)
```
Use the type `RRefForkData` here instead of IValue.

pritamdamania87 commented on torch/csrc/distributed/rpc/script_rref_proto.h:
```diff
@@ -0,0 +1,70 @@
+// OwnerRRef uses this message to send the RRef value to a remote UserRRef
+class TORCH_API ScriptRRefFetchRet final : public RRefMessageBase {
+ public:
+  ScriptRRefFetchRet(at::IValue value)
```
All of these APIs should use `RRefForkData` instead of generic IValue.

pritamdamania87 commented on torch/csrc/distributed/rpc/types.cpp:
```diff
@@ -0,0 +1,54 @@
+      createdOn);
+
+  TORCH_CHECK(
+      localId < std::numeric_limits<local_id_t>::max(),
```
The localId actually has a lower max since it can use only 48 bits.

mrshenli replied:
In this PR `GloballyUniqueId` is a struct of two numbers, and the 48 bits is only applied when calculating hash. We can add those when we consolidate this with autograd context id.

pritamdamania87 commented on torch/csrc/distributed/rpc/types.cpp:
```diff
@@ -0,0 +1,54 @@
+  local_id_t localId = ivalues[1].toInt();
+
+  TORCH_CHECK(
+      createdOn < std::numeric_limits<worker_id_t>::max(),
```
Why not `<=`?

pritamdamania87 commented on torch/csrc/distributed/rpc/types.h:
```diff
@@ -0,0 +1,42 @@
+using worker_id_t = int16_t;
+using local_id_t = uint64_t;
+
+struct GloballyUniqueId final {
```
Can we use `operator int()` to implicitly convert this to int64_t? This way we don't need to worry about Hash etc. and can easily use this object as an int.

pritamdamania87 commented on test/test_rpc.py:
```diff
@@ -385,5 +385,33 @@ def test_stress_light_rpc(self):
+            expected.append(torch.ones(n, n) * 2)
+
+        for i in range(m):
+            self.assertEqual(rrefs[i].to_here(), expected[i])
```
We need some more tests here:

1. Verify all RRefs are cleaned up properly from RRefContext in the end.
2. Validate state of RRefContext on each node after we've passed a bunch of RRef's around. Also, validate forks are created appropriately.
3. Add test to verify owner of RRef is populated correctly.
4. Take down the owner of an rref and verify to_here() throws an appropriate exception.

mrshenli replied:
> Take down the owner of an rref and verify to_here() throws an appropriate exception.

@pritamdamania87  How do we take down the owner of an RRef? Are we going to expose test-only APIs for this?

pritamdamania87 commented:
@aazzolini Thanks for reviewing the PR, although I don't think this PR was in any state to be landed. I got a chance to review it recently and have a lot of concerns as you can see inline. At a high level, the PR needed:

1) More test coverage
2) Needed much more documentation.
3) Some of the APIs needed to be re-worked. 
4) Needed a lot more validation in certain places.
5) We could have thrown exceptions in destructors which is very dangerous.

All of these issues should have been addressed in this PR and not in a PR on top of this.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -0,0 +1,69 @@
+
+std::unique_ptr<RRefContext> RRefContext::context_;
+
+void RRefContext::initInstance(std::shared_ptr<RpcAgent> agent) {
```
`RRefContext` need to send several RRef reference counting related messages, and hence need access to the `RpcAgent`.

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -31,30 +66,46 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
-      kwargs,
-      ") to a builtin operator");
+  auto& ctx = RRefContext::getInstance();
+  if (ctx->getWorkerId() == dst.id_) {
```
Yes, users might want to create an `OwnerRRef` on the return value of a self-RPC call. After that, they can immediately pass the RRef to others withut waiting for the self-RPC to finish.

mrshenli commented on torch/csrc/distributed/rpc/types.h:
```diff
@@ -0,0 +1,42 @@
+using worker_id_t = int16_t;
+using local_id_t = uint64_t;
+
+struct GloballyUniqueId final {
```
Yes, let's put that in some common folder.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,110 @@
+namespace distributed {
+namespace rpc {
+
+// Manages RRef lifetime and keeps track of RRef forks.
```
Yes, it is a singleton. I will add comments here. Most comments regarding lifetime are in `RRef.h` in the PR on top of it.

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -63,9 +41,67 @@ std::shared_ptr<FutureMessage> py_rpc_builtin(
+  auto op = matchBuiltinOp(opName, args, kwargs, stack);
+
+  auto& ctx = RRefContext::getInstance();
+  auto userRRef = ctx->createUserRRef(dst.id_);
```
Yes, that will make the reference counting algorithm cleaner, but one downside is that the it has to wait for a round-trip before the caller can share the RRef with anyone else.

mrshenli commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +61,50 @@ Message processRequestBlocking(Message&& request) {
+
+      auto rrefId = RRefId::fromIValue(src.retRRefId());
+      auto forkId = ForkId::fromIValue(src.retForkId());
+      TORCH_CHECK(rrefId != forkId, "Does not support remote call to self.");
```
They are the same when making RPC to self, where the caller also holds the owner, and owner does not have a forkID.

mrshenli commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +61,50 @@ Message processRequestBlocking(Message&& request) {
+          "size ",
+          stack.size());
+
+      ownerRRef->setValue(std::move(stack.front()));
```
this + not creating RRef on caller side will force the caller to wait a round-trip + the RPC execution time, which can be significant.

mrshenli commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +61,50 @@ Message processRequestBlocking(Message&& request) {
+      ScriptRRefFetchCall srf = ScriptRRefFetchCall::fromMessage(request);
+      // TODO: make this asynchronous
+      std::shared_ptr<OwnerRRef<IValue>> rref =
+          RRefContext::getInstance()->getOrCreateOwnerRRef<IValue>(
```
This is not a bug if we don't enforce `OwnerRRef` be created before any `UserRRef`, because the caller could have shared the `UserRRef` with other users, and those users can communicate with the owner before the caller's message arrives at the owner. 

Let's imagine a model parallel training case. Say there is a trainer that drives the training, and the model locates on 10 machines. One implementation could be the trainer make a remote call to the machine that hosts the first piece of the model, and immediately gets the RRef of the return value, then forward the RRef to the second piece of the model, and so on. If we enforce creating `OwnerRRef` before `UserRRef`, then we are forcing 10 more round-trips delay in one iteration. If we create `OwnerRRef` after the operator execution is done, we are then forcing the model to be run in a serialized manner, while the model could be two tiers of 5 parallel layers. 

It is true that the exception of those operation execution hasn't been captured properly, and they should. But I feel we don't have to create `OwnerRRef` after the operation finishes, because if there are failures, the one who retrieves the value from the `RRef` will see that error. I will add this. 

Exclude the exception handling part, then the question is do we want to trade performance for code simplicity?

mrshenli commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +61,50 @@ Message processRequestBlocking(Message&& request) {
+    }
+    case MessageType::RREF_FETCH_CALL: {
+      ScriptRRefFetchCall srf = ScriptRRefFetchCall::fromMessage(request);
+      // TODO: make this asynchronous
```
Yes, let me add issues. It can register callbacks on OwnerRRef, which will be fired when the value is ready.

mrshenli commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,140 @@
+  friend class UserRRef;
+
+  RRefForkData(
+      worker_id_t ownerId,
```
It is necessary when RRefs are shared between users.

mrshenli commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,140 @@
+
+// Represents fork of an RRef to be sent over the wire.
+//
+// In order to preserve correctness of reference counting, each RRefForkData
```
Yes, they are enforced at a higher layer, but anyone using `RRefForkData` should be aware. They should not create multiple RRefs from `RRefForkData` (this can be done through `RRefContext`), and they should not throw away `RRefForkData` they got from `fork()` (because the user could have notified the owner on the fork).

mrshenli commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,140 @@
+    C10_IS_TRIVIALLY_COPYABLE(RRefForkData),
+    "RRefForkData must be trivially copyable");
+
+// TODO: make RRef an IValue, and edit createStackForSchema accordingly
```
Docs are added in the PR on top of this. This is not the full-fledged RRef, so didn't spend time to describe details on which part of RRef is implemented in this PR.

mrshenli replied:
For IValue, I was thinking about using RRef as args for builtin ops, it will be nice if we can make `createStackForSchema` create the stack for us.

mrshenli commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,140 @@
+  ~UserRRef() override;
+
+ private:
+  friend class RRefContext;
```
RRef/UserRRef/OwnerRRef will be public C++ APIs, but applications have to create them through `RRefContext`. So used the friend class to access the private constructor.

mrshenli commented on torch/csrc/distributed/rpc/rref.h:
```diff
@@ -0,0 +1,140 @@
+    return value_.value();
+  }
+
+  void setValue(T&& value) {
```
The associated value could take time to compute, and we would like to support sharing RRefs before the value is ready.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -0,0 +1,72 @@
+
+void RRefContext::addFork(const at::IValue& value) {
+  auto rfd = RRefForkData::fromIValue(value);
+  AT_ASSERT(
```
no, should have used `TORCH_INTERNAL_ASSERT `

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -0,0 +1,72 @@
+      rfd.ownerId_ == getWorkerId(),
+      "RRef user should never receive fork notification.");
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto& rrefForks = forks_[rfd.rrefId_];
```
owners might not be here yet, if we don't enforce creating owner before users

mrshenli commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -0,0 +1,72 @@
+      rfd.ownerId_ == getWorkerId(),
+      "RRef user should never receive delete notification.");
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto& rrefForks = forks_[rfd.rrefId_];
```
yes, let me add that.

mrshenli commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,116 @@
+  const std::shared_ptr<RpcAgent> agent_;
+  std::mutex mutex_;
+  // Keep OwnerRRefs alive until there is no living UserRRefs.
+  std::unordered_map<RRefId, std::shared_ptr<RRef>, RRefId::Hash> owners_;
```
cause under the current design (add complexity to make RRef sharing fast), owners might have not be created yet.

gqchen replied:
can we be explicit here? use `std::share_ptr<OwnerRRef>`, instead of `std::share_ptr<RRef>`

mrshenli commented on torch/csrc/distributed/rpc/script_call.cpp:
```diff
@@ -21,14 +21,16 @@ const std::vector<at::IValue>& ScriptCall::stack() const {
 
-Message ScriptCall::toMessage() {
-  std::vector<IValue> ivalues;
+std::vector<at::IValue>& ScriptCall::stackRef() {
```
I saw jit use `toStringRef`, do we have a convention here?

mrshenli commented on torch/csrc/distributed/rpc/script_remote_call.cpp:
```diff
@@ -0,0 +1,60 @@
+ScriptRemoteCall::ScriptRemoteCall(
+    std::shared_ptr<Operator> op,
+    std::vector<at::IValue>&& args,
+    at::IValue retRRefId,
```
I see this as one of the main discussion we should do offline: performance vs complexity

aazzolini commented:
@pritamdamania87 you're right, my accept was hasty. @mrshenli, definitely address request from other reviewers before moving on to #25499.

gqchen commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,116 @@
+
+  // create a new RRef
+  template <typename T>
+  std::shared_ptr<OwnerRRef<T>> createOwnerRRef(worker_id_t ownerId) {
```
don't we want some lightweight functions to be inlined?

gqchen commented on torch/csrc/distributed/rpc/rref_context.h:
```diff
@@ -0,0 +1,116 @@
+    if (ownerId == getWorkerId()) {
+      return getOrCreateOwnerRRef<T>(rrefId);
+    } else {
+      return createUserRRef(ownerId, rrefId, forkId);
```
should we separate out getOrCreateOwnerRRef() and getOrCreateUserRRef()? Do we need getOrCreateUserRRef?

gqchen commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +61,50 @@ Message processRequestBlocking(Message&& request) {
+      std::shared_ptr<OwnerRRef<IValue>> rref =
+          RRefContext::getInstance()->getOrCreateOwnerRRef<IValue>(
+              RRefId::fromIValue(srf.value()));
+      auto response = ScriptRRefFetchRet(rref->getValue()).toMessage();
```
if value is not available, this will block forever. We should add a timeout here.

gqchen commented on torch/csrc/distributed/rpc/functions.cpp:
```diff
@@ -50,6 +61,50 @@ Message processRequestBlocking(Message&& request) {
+      TORCH_CHECK(rrefId != forkId, "Does not support remote call to self.");
+
+      auto& ctx = RRefContext::getInstance();
+      auto ownerRRef = ctx->getOrCreateOwnerRRef<IValue>(rrefId);
```
we can addFork here?

gqchen reviewed (commented):
a few more issues. I agree with Pritam that we need to get this PR wrapped up before moving on to add new features.
